osu.Game/Beatmaps/Formats/OsuLegacyDecoder.cs

```diff
@@ -0,0 +1,272 @@
+            // Note: the combo index specified in the beatmap is discarded
+            beatmap.ComboColors.Add(new Color4
+            {
+                R = r / 255f,
+                G = g / 255f,
+                B = b / 255f,
+                A = 1f,
+            });
+        }
+
+        public override Beatmap Decode(TextReader stream)
+        {
+            var beatmap = new Beatmap
+            {
+                Metadata = new BeatmapMetadata(),
+                BaseDifficulty = new BaseDifficulty(),
+                HitObjects = new List<HitObject>(),
+                ControlPoints = new List<ControlPoint>(),
+                ComboColors = new List<Color4>(),
+            };
+            var section = Section.None;
+            string line;
+            while (true)
+            {
+                line = stream.ReadLine();
+                if (line == null)
+                    break;
+                line = line.Trim();
+                if (string.IsNullOrEmpty(line))
+                    continue;
+                if (line.StartsWith(@"osu file format v"))
+                    continue;
+                    
+                if (line.StartsWith(@"[") && line.EndsWith(@"]"))
+                {
+                    if (!Enum.TryParse(line.Substring(1, line.Length - 2), out section))
+                        throw new InvalidDataException($@"Unknown osu section {line}");
+                    continue;
+                }
+                
+                string val = line, key = null;
+                if (section != Section.Events && section != Section.TimingPoints && section != Section.HitObjects)
+                {
+                    key = val.Remove(val.IndexOf(':')).Trim();
+                    val = val.Substring(val.IndexOf(':') + 1).Trim();
+                }
+                switch (section)
+                {
+                    case Section.General:
+                        handleGeneral(beatmap, key, val);
+                        break;
+                    case Section.Editor:
+                        handleEditor(beatmap, key, val);
+                        break;
+                    case Section.Metadata:
+                        handleMetadata(beatmap, key, val);
+                        break;
+                    case Section.Difficulty:
+                        handleDifficulty(beatmap, key, val);
+                        break;
+                    case Section.Events:
+                        handleEvents(beatmap, val);
+                        break;
+                    case Section.TimingPoints:
+                        handleTimingPoints(beatmap, val);
+                        break;
+                    case Section.Colours:
+                        handleColours(beatmap, key, val);
+                        break;
+                    case Section.HitObjects:
+                        beatmap.HitObjects.Add(HitObject.Parse(beatmap.Mode, val));

```

Comment:
not too fond of having this the way it is. i'm fine to merge it for this PR, but can we agree on having something better like

```
var parser = HitObjectParser.ForMode(beatmap.Mode);

...
for (var line in file)
    parser.Parse(line);
```